File: sem iii/NM/Nm lab/ho/NewtonRaphson_lab.py
import math

# Lab 2 - Newton Raphson Method
# f(x) = x^3 - 2x - 5

# define function f(x)
def f(x):
    return x**3 - 2*x - 5

# define derivative f'(x)
def fd(x):
    return 3*x**2 - 2


# Newton Raphson function
def NewtonRaphson(x, itr, eps):
    for i in range(itr):
        fx = f(x)
        fdx = fd(x)

        # Newton-Raphson formula
        x_new = x - (fx / fdx)

        # relative approximate error
        error = abs((x_new - x) / x_new)

        print(f"The root estimate is: {x_new:.5f} and the error is: {error:.5f}")

        # stopping condition
        if error < eps:
            return x_new

        # update x
        x = x_new

    return x


if __name__ == '__main__':
    print("Program to find the root of the given function using Newton Raphson method.")
    x = float(input("Enter the initial guess: "))
    eps = float(input("Enter the error tolerance value: "))
    itr = int(input("Enter the number of iterations: "))

    root = NewtonRaphson(x, itr, eps)
    print(f"\nThe root of the function is {root:.5f}")

File: sem iii/NM/Nm lab/ho/NewtonRaphson_lab.c
#include <stdio.h>
#include <math.h>

// f(x) = x^3 - 2x - 5
double f(double x) {
    return x*x*x - 2*x - 5;
}

// f'(x) = 3x^2 - 2
double fd(double x) {
    return 3*x*x - 2;
}

int main() {
    double x, x_new, error, eps;
    int itr, i;

    printf("Program to find root using Newton Raphson Method\n");
    printf("Enter the initial guess: ");
    scanf("%lf", &x);
    printf("Enter the error tolerance: ");
    scanf("%lf", &eps);
    printf("Enter the number of iterations: ");
    scanf("%d", &itr);

    for (i = 0; i < itr; i++) {
        x_new = x - (f(x) / fd(x));
        error = fabs((x_new - x) / x_new);

        printf("Iteration %d: Root = %.5lf  Error = %.5lf\n",
               i+1, x_new, error);

        if (error < eps)
            break;

        x = x_new;
    }

    printf("\nThe root of the function is %.5lf\n", x_new);
    return 0;
}
